File: dotnet/src/EPostak/Resources/EventsResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>Preferred pull/ack event facade over the webhook queue.</summary>
public sealed class EventsResource
{
    private readonly HttpRequestor _http;

    internal EventsResource(HttpRequestor http) => _http = http;

    public Task<WebhookQueueResponse> PullAsync(WebhookQueueParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("limit", @params?.Limit?.ToString()),
            ("event_type", @params?.EventType));
        return _http.RequestAsync<WebhookQueueResponse>(HttpMethod.Get, $"/events/pull{qs}", ct);
    }

    public Task<AckResponse> AckAsync(string eventId, CancellationToken ct = default)
        => _http.RequestAsync<AckResponse>(HttpMethod.Post, $"/events/{Uri.EscapeDataString(eventId)}/ack", new { }, ct);

    public Task<BatchAckResponse> BatchAckAsync(IEnumerable<string> eventIds, CancellationToken ct = default)
        => _http.RequestAsync<BatchAckResponse>(HttpMethod.Post, "/events/batch-ack", new { event_ids = eventIds }, ct);
}

File: java/src/main/java/sk/epostak/sdk/resources/EventsResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.AckResponse;
import sk.epostak.sdk.models.BatchAckResponse;
import sk.epostak.sdk.models.WebhookQueueResponse;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Preferred pull/ack event facade over the webhook queue. */
public final class EventsResource {

    private final HttpClient http;

    public EventsResource(HttpClient http) {
        this.http = http;
    }

    public WebhookQueueResponse pull(Integer limit, String eventType) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("limit", limit);
        params.put("event_type", eventType);
        return http.get("/events/pull" + HttpClient.buildQuery(params), WebhookQueueResponse.class);
    }

    public WebhookQueueResponse pull() {
        return pull(null, null);
    }

    public AckResponse ack(String eventId) {
        return http.post("/events/" + HttpClient.encode(eventId) + "/ack", Map.of(), AckResponse.class);
    }

    public BatchAckResponse batchAck(List<String> eventIds) {
        return http.post("/events/batch-ack", Map.of("event_ids", eventIds), BatchAckResponse.class);
    }
}
